File: Crio/crio_programming_interview_problems-master/WallsAndGates/WallsAndGates.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> wallsAndGates(vector<vector<int>> &rooms)
{
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    vector<int> dir = {-1, 0, 1, 0, -1};
    queue<pair<int, int>> q;

    for (int i = 0; i < rooms.size(); i++)
    {
        for (int j = 0; j < rooms[i].size(); j++)
        {
            if (rooms[i][j] == 0)
            {
                q.push(make_pair(i, j));
            }
        }
    }
    while (!q.empty())
    {
        int n = rooms.size();
        int m = rooms[0].size();

        pair<int, int> top = q.front();
        q.pop();

        for (int i = 0; i < 4; i++)
        {
            int r = top.first + dir[i];
            int c = top.second + dir[i + 1];
            if (r >= 0 && r < n && c >= 0 && c < m && rooms[r][c] == INT_MAX)
            {
                rooms[r][c] = rooms[top.first][top.second] + 1;
                q.push(make_pair(r, c));
            }
        }
    }
    return rooms;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

int main()
{
    vector<vector<int>> grid;
    int n, m;
    cin >> n >> m;

    grid.resize(n);
    for (int i = 0; i < n; i++)
    {
        vector<int> b(m, 0);
        for (int j = 0; j < m; j++)
        {
            cin >> b[j];
        }
        grid[i] = b;
    }
    vector<vector<int>> ans = wallsAndGates(grid);
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Crio/crio_programming_interview_problems-master/WallsAndGates/WallsAndGates.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
from queue import Queue
INF = 2147483647
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def wallsAndGates(n,m,grid):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    dis = [[INF]*m for _ in range(n)]
    q = Queue(maxsize=0)
    for i in range(n):
        for j in range(m):
            if(grid[i][j]==0):
                q.put([i,j])
                dis[i][j]=0
            if(grid[i][j]==-1):
                dis[i][j]=-1
    while(not q.empty()):
        i,j = q.get()
        for r,c in [[i-1,j], [i+1,j], [i,j-1], [i,j+1]]:
            if(r>=0 and r<n and c>=0 and c<m):
                if(dis[r][c] > dis[i][j]+1):
                    dis[r][c] = dis[i][j]+1
                    q.put([r,c])

    return dis
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

if __name__=="__main__":
    n,m = map(int,input().split())
    adj = []
    for i in range(n):
        adj.append(list(map(int,input().split())))
    ans = wallsAndGates(n,m,adj)
    for row in ans:
        print(*row)
